File: frontend/src/filterEngine.js
function resizeToCanvas(image, width, height) {
  const canvas = document.createElement("canvas");
  canvas.width = width;
  canvas.height = height;
  canvas.getContext("2d").drawImage(image, 0, 0, width, height);
  return canvas;
}

function rotateImageWithAnchor(sourceCanvas, angleDeg, anchorX, anchorY) {
  const width = sourceCanvas.width;
  const height = sourceCanvas.height;
  const radians = (angleDeg * Math.PI) / 180;
  const cos = Math.abs(Math.cos(radians));
  const sin = Math.abs(Math.sin(radians));
  const newWidth = Math.ceil(height * sin + width * cos);
  const newHeight = Math.ceil(height * cos + width * sin);

  const canvas = document.createElement("canvas");
  canvas.width = newWidth;
  canvas.height = newHeight;
  const ctx = canvas.getContext("2d");

  const centerX = width / 2;
  const centerY = height / 2;

  ctx.translate(newWidth / 2, newHeight / 2);
  ctx.rotate(radians);
  ctx.drawImage(sourceCanvas, -centerX, -centerY);

  const dx = anchorX - centerX;
  const dy = anchorY - centerY;
  const mappedX =
    newWidth / 2 + dx * Math.cos(radians) - dy * Math.sin(radians);
  const mappedY =
    newHeight / 2 + dx * Math.sin(radians) + dy * Math.cos(radians);

  return { canvas, anchorX: mappedX, anchorY: mappedY };
}

function overlayCanvas(targetCtx, overlayCanvas, x, y) {
  targetCtx.drawImage(overlayCanvas, Math.round(x), Math.round(y));
}

function placeRotatedFilter(
  targetCtx,
  filterImage,
  width,
  height,
  angle,
  targetX,
  targetY,
  anchorX,
  anchorY
) {
  const resized = resizeToCanvas(filterImage, width, height);
  const { canvas, anchorX: mappedX, anchorY: mappedY } = rotateImageWithAnchor(
    resized,
    angle,
    anchorX,
    anchorY
  );

  overlayCanvas(targetCtx, canvas, targetX - mappedX, targetY - mappedY);
}

function applyGlasses(ctx, landmarks, filterImage) {
  const [leftOuterX, leftOuterY] = landmarks.left_eye_outer;
  const [leftInnerX, leftInnerY] = landmarks.left_eye_inner;
  const [rightInnerX, rightInnerY] = landmarks.right_eye_inner;
  const [rightOuterX, rightOuterY] = landmarks.right_eye_outer;

  const leftEyeX = Math.floor((leftOuterX + leftInnerX) / 2);
  const leftEyeY = Math.floor((leftOuterY + leftInnerY) / 2);
  const rightEyeX = Math.floor((rightInnerX + rightOuterX) / 2);
  const rightEyeY = Math.floor((rightInnerY + rightOuterY) / 2);

  const eyeCenterX = Math.floor((leftEyeX + rightEyeX) / 2);
  const eyeCenterY = Math.floor((leftEyeY + rightEyeY) / 2);
  const eyeDistance = Math.floor(
    Math.hypot(rightEyeX - leftEyeX, rightEyeY - leftEyeY)
  );
  const angle =
    -(Math.atan2(leftEyeY - rightEyeY, rightEyeX - leftEyeX) * 180) / Math.PI;

  const glassesWidth = Math.floor(eyeDistance * 2.25);
  const aspectRatio = filterImage.naturalHeight / filterImage.naturalWidth;
  const glassesHeight = Math.floor(glassesWidth * aspectRatio);

  placeRotatedFilter(
    ctx,
    filterImage,
    glassesWidth,
    glassesHeight,
    angle,
    eyeCenterX,
    eyeCenterY,
    Math.floor(glassesWidth * 0.5),
    Math.floor(glassesHeight * 0.38)
  );
}

function applyMoustache(ctx, landmarks, filterImage) {
  const [x1, y1] = landmarks.left_mouth;
  const [x2, y2] = landmarks.right_mouth;
  const [, ny] = landmarks.nose_tip;

  const mouthWidth = Math.floor(Math.hypot(x2 - x1, y2 - y1));
  const angle = -(Math.atan2(y1 - y2, x2 - x1) * 180) / Math.PI;
  const targetX = Math.floor((x1 + x2) / 2);
  const targetY = Math.floor((ny + Math.floor((y1 + y2) / 2)) / 2) + 5;

  const moustacheWidth = Math.floor(mouthWidth * 1.4);
  const aspectRatio = filterImage.naturalHeight / filterImage.naturalWidth;
  const moustacheHeight = Math.floor(moustacheWidth * aspectRatio);

  placeRotatedFilter(
    ctx,
    filterImage,
    moustacheWidth,
    moustacheHeight,
    angle,
    targetX,
    targetY,
    Math.floor(moustacheWidth / 2),
    Math.floor(moustacheHeight / 2)
  );
}

function applyCrown(ctx, landmarks, filterImage) {
  const [x1, y1] = landmarks.left_face;
  const [x2, y2] = landmarks.right_face;
  const [fx, fy] = landmarks.forehead;

  const faceWidth = Math.floor(Math.hypot(x2 - x1, y2 - y1));
  const angle = -(Math.atan2(y1 - y2, x2 - x1) * 180) / Math.PI;
  const crownWidth = Math.floor(faceWidth * 1.2);
  const aspectRatio = filterImage.naturalHeight / filterImage.naturalWidth;
  const crownHeight = Math.floor(crownWidth * aspectRatio);

  placeRotatedFilter(
    ctx,
    filterImage,
    crownWidth,
    crownHeight,
    angle,
    fx,
    fy + 8,
    Math.floor(crownWidth / 2),
    crownHeight
  );
}

function applyCowboyHat(ctx, landmarks, filterImage) {
  const [x1, y1] = landmarks.left_face;
  const [x2, y2] = landmarks.right_face;
  const [fx, fy] = landmarks.forehead;

  const faceWidth = Math.floor(Math.hypot(x2 - x1, y2 - y1));
  const angle = -(Math.atan2(y1 - y2, x2 - x1) * 180) / Math.PI;
  const hatWidth = Math.floor(faceWidth * 1.5);
  const aspectRatio = filterImage.naturalHeight / filterImage.naturalWidth;
  const hatHeight = Math.floor(hatWidth * aspectRatio);

  placeRotatedFilter(
    ctx,
    filterImage,
    hatWidth,
    hatHeight,
    angle,
    fx,
    fy + 48,
    Math.floor(hatWidth / 2),
    hatHeight
  );
}

function applyBowTie(ctx, landmarks, filterImage) {
  const [x1, y1] = landmarks.left_mouth;
  const [x2, y2] = landmarks.right_mouth;
  const [cx, cy] = landmarks.chin;

  const tieWidth = Math.floor(Math.hypot(x2 - x1, y2 - y1) * 2);
  const angle = -(Math.atan2(y1 - y2, x2 - x1) * 180) / Math.PI;
  const aspectRatio = filterImage.naturalHeight / filterImage.naturalWidth;
  const tieHeight = Math.floor(tieWidth * aspectRatio);

  placeRotatedFilter(
    ctx,
    filterImage,
    tieWidth,
    tieHeight,
    angle,
    cx,
    cy + Math.floor(tieHeight * 0.4),
    Math.floor(tieWidth / 2),
    Math.floor(tieHeight / 2)
  );
}

function applyMood(ctx, landmarks, filterImages) {
  const [, leftMouthY] = landmarks.left_mouth;
  const [, rightMouthY] = landmarks.right_mouth;
  const [, topLipY] = landmarks.top_lip;
  const [, bottomLipY] = landmarks.bottom_lip;

  const cornerY = (leftMouthY + rightMouthY) / 2;
  const lipY = (topLipY + bottomLipY) / 2;

  // If corners are higher than lip center, it's a smile (Y goes down, so smaller Y is higher)
  const isHappy = cornerY < lipY;
  const filterImage = isHappy ? filterImages[0] : filterImages[1];

  const [lx, ly] = landmarks.left_face;
  const [rx, ry] = landmarks.right_face;

  const faceWidth = Math.floor(Math.hypot(rx - lx, ry - ly));
  
  const emojiWidth = Math.floor(faceWidth * 0.4);
  const aspectRatio = filterImage.naturalHeight / filterImage.naturalWidth;
  const emojiHeight = Math.floor(emojiWidth * aspectRatio);

  const targetX = rx + Math.floor(emojiWidth * 0.7);
  const targetY = ry - Math.floor(faceWidth * 0.2);

  // Negative angle in Canvas API is counter-clockwise (left tilt)
  const angle = -15;

  placeRotatedFilter(
    ctx,
    filterImage,
    emojiWidth,
    emojiHeight,
    angle,
    targetX,
    targetY,
    Math.floor(emojiWidth / 2),
    Math.floor(emojiHeight / 2)
  );
}

const FILTER_APPLIERS = {
  Glasses: applyGlasses,
  Moustache: applyMoustache,
  Crown: applyCrown,
  "Cowboy Hat": applyCowboyHat,
  "Bow Tie": applyBowTie,
  Mood: applyMood,
};

export function renderFilteredImage(baseImage, filterImage, filterName, landmarks) {
  const canvas = document.createElement("canvas");
  canvas.width = baseImage.naturalWidth || baseImage.width;
  canvas.height = baseImage.naturalHeight || baseImage.height;

  const ctx = canvas.getContext("2d");
  ctx.drawImage(baseImage, 0, 0, canvas.width, canvas.height);

  const applyFilter = FILTER_APPLIERS[filterName];
  if (applyFilter && landmarks) {
    applyFilter(ctx, landmarks, filterImage);
  }

  return canvas;
}

File: backend/filter_engine.py
import cv2
import numpy as np
from utils import overlay_png, rotate_image_with_anchor


class FilterEngine:
    def __init__(self):
        pass

    def apply_glasses(self, image, landmarks, glasses_path):
        if landmarks is None:
            return image

        output = image.copy()

        left_outer = landmarks["left_eye_outer"]
        left_inner = landmarks["left_eye_inner"]
        right_inner = landmarks["right_eye_inner"]
        right_outer = landmarks["right_eye_outer"]

        left_eye_x = (left_outer[0] + left_inner[0]) // 2
        left_eye_y = (left_outer[1] + left_inner[1]) // 2

        right_eye_x = (right_inner[0] + right_outer[0]) // 2
        right_eye_y = (right_inner[1] + right_outer[1]) // 2

        eye_center_x = (left_eye_x + right_eye_x) // 2
        eye_center_y = (left_eye_y + right_eye_y) // 2

        eye_distance = int(np.sqrt((right_eye_x - left_eye_x) ** 2 +
                                   (right_eye_y - left_eye_y) ** 2))

        angle = np.degrees(np.arctan2(left_eye_y - right_eye_y,
                                      right_eye_x - left_eye_x))

        glasses = cv2.imread(glasses_path, cv2.IMREAD_UNCHANGED)
        if glasses is None:
            raise FileNotFoundError(glasses_path)

        glasses_width = int(eye_distance * 2.25)
        aspect_ratio = glasses.shape[0] / glasses.shape[1]
        glasses_height = int(glasses_width * aspect_ratio)
        glasses = cv2.resize(glasses, (glasses_width, glasses_height))

        bridge_anchor = (
            int(glasses.shape[1] * 0.50),
            int(glasses.shape[0] * 0.38)
        )

        rotated, (ax, ay) = rotate_image_with_anchor(glasses, angle, bridge_anchor)

        top_left_x = int(eye_center_x - ax)
        top_left_y = int(eye_center_y - ay)

        return overlay_png(output, rotated, top_left_x, top_left_y)

    def apply_moustache(self, image, landmarks, moustache_path):
        if landmarks is None:
            return image

        left_mouth = landmarks["left_mouth"]
        right_mouth = landmarks["right_mouth"]
        nose = landmarks["nose_tip"]

        x1, y1 = left_mouth
        x2, y2 = right_mouth
        nx, ny = nose

        mouth_width = int(np.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2))
        angle = np.degrees(np.arctan2(y1 - y2, x2 - x1))

        moustache = cv2.imread(moustache_path, cv2.IMREAD_UNCHANGED)
        if moustache is None:
            raise FileNotFoundError(moustache_path)

        moustache_width = int(mouth_width * 1.4)
        aspect_ratio = moustache.shape[0] / moustache.shape[1]
        moustache_height = int(moustache_width * aspect_ratio)
        moustache = cv2.resize(moustache, (moustache_width, moustache_height))

        target_x = (x1 + x2) // 2
        target_y = (ny + (y1 + y2) // 2) // 2 + 5

        anchor = (moustache.shape[1] // 2, moustache.shape[0] // 2)

        rotated, (ax, ay) = rotate_image_with_anchor(moustache, angle, anchor)

        top_left_x = int(target_x - ax)
        top_left_y = int(target_y - ay)

        return overlay_png(image.copy(), rotated, top_left_x, top_left_y)

    def apply_crown(self, image, landmarks, crown_path):
        if landmarks is None:
            return image

        left_face = landmarks["left_face"]
        right_face = landmarks["right_face"]
        forehead = landmarks["forehead"]

        x1, y1 = left_face
        x2, y2 = right_face
        fx, fy = forehead

        face_width = int(np.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2))
        angle = np.degrees(np.arctan2(y1 - y2, x2 - x1))

        crown = cv2.imread(crown_path, cv2.IMREAD_UNCHANGED)
        if crown is None:
            raise FileNotFoundError(crown_path)

        crown_width = int(face_width * 1.2)
        aspect_ratio = crown.shape[0] / crown.shape[1]
        crown_height = int(crown_width * aspect_ratio)
        crown = cv2.resize(crown, (crown_width, crown_height))

        target_x = fx
        target_y = fy + 8

        anchor = (crown.shape[1] // 2, crown.shape[0])

        rotated, (ax, ay) = rotate_image_with_anchor(crown, angle, anchor)

        top_left_x = int(target_x - ax)
        top_left_y = int(target_y - ay)

        return overlay_png(image.copy(), rotated, top_left_x, top_left_y)

    def apply_cowboy_hat(self, image, landmarks, cowboy_hat_path):
        if landmarks is None:
            return image

        left_face = landmarks["left_face"]
        right_face = landmarks["right_face"]
        forehead = landmarks["forehead"]

        x1, y1 = left_face
        x2, y2 = right_face
        fx, fy = forehead

        face_width = int(np.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2))
        angle = np.degrees(np.arctan2(y1 - y2, x2 - x1))

        hat = cv2.imread(cowboy_hat_path, cv2.IMREAD_UNCHANGED)
        if hat is None:
            raise FileNotFoundError(cowboy_hat_path)

        hat_width = int(face_width * 1.5)
        aspect_ratio = hat.shape[0] / hat.shape[1]
        hat_height = int(hat_width * aspect_ratio)
        hat = cv2.resize(hat, (hat_width, hat_height))

        target_x = fx
        target_y = fy + 48

        anchor = (hat.shape[1] // 2, hat.shape[0])

        rotated, (ax, ay) = rotate_image_with_anchor(hat, angle, anchor)

        top_left_x = int(target_x - ax)
        top_left_y = int(target_y - ay)

        return overlay_png(image.copy(), rotated, top_left_x, top_left_y)

    def apply_bow_tie(self, image, landmarks, bow_tie_path):
        if landmarks is None:
            return image

        left_mouth = landmarks["left_mouth"]
        right_mouth = landmarks["right_mouth"]
        chin = landmarks["chin"]

        x1, y1 = left_mouth
        x2, y2 = right_mouth
        cx, cy = chin

        tie_width = int(np.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2) * 2.0)
        angle = np.degrees(np.arctan2(y1 - y2, x2 - x1))

        tie = cv2.imread(bow_tie_path, cv2.IMREAD_UNCHANGED)
        if tie is None:
            raise FileNotFoundError(bow_tie_path)

        aspect_ratio = tie.shape[0] / tie.shape[1]
        tie_height = int(tie_width * aspect_ratio)
        tie = cv2.resize(tie, (tie_width, tie_height))

        target_x = cx
        target_y = cy + int(tie_height * 0.4)

        anchor = (tie.shape[1] // 2, tie.shape[0] // 2)

        rotated, (ax, ay) = rotate_image_with_anchor(tie, angle, anchor)

        top_left_x = int(target_x - ax)
        top_left_y = int(target_y - ay)

        return overlay_png(image.copy(), rotated, top_left_x, top_left_y)